src/entities/trade.ts

```diff
@@ -576,4 +579,77 @@ export class Trade {
     }
     return methodName
   }
+
+  /**
+   * return the mistX router method name for the trade
+   * @param pairs
+   * @param currencyIn 
+   * @param currencyOut 
+   * @param gasPriceToBeat 
+   * @param minerBribeMargin
+   * @param maxHops maximum number of hops a returned trade can make, e.g. 1 hop goes through a single pair
+   */
+   public static estimateMinTradeAmounts(
+    pairs: Pair[],
+    currencyIn: Currency,
+    currencyOut: Currency, 
+    gasPriceToBeat: BigintIsh,
+    minerBribeMargin: BigintIsh,
+    minTradeMargin: BigintIsh,
+    { maxHops = 3 }: BestTradeOptions = {}
+  ): MinTradeEstimate | null {
+    const etherIn = currencyIn === ETHER
+    const etherOut = currencyOut === ETHER
+
+    if (!etherIn && !etherOut) return null;
+
+    const exactInGas = estimatedGasForMethod(Trade.methodNameForTradeType(TradeType.EXACT_INPUT, etherIn, etherOut), maxHops.toString())
+    const exactOutGas = estimatedGasForMethod(Trade.methodNameForTradeType(TradeType.EXACT_OUTPUT, etherIn, etherOut), maxHops.toString())
+
+    const exactInBribe = calculateMargin(calculateMinerBribe(gasPriceToBeat, exactInGas, minerBribeMargin), minTradeMargin)
+    const exactOutBribe = calculateMargin(calculateMinerBribe(gasPriceToBeat, exactOutGas, minerBribeMargin), minTradeMargin)
+
+    const chainId: ChainId | undefined = (currencyIn as Token).chainId || (currencyOut as Token).chainId || undefined
+    invariant(chainId, 'BRIBE_ESTIMATES_CHAINID')
+    let amountOut: TokenAmount = wrappedAmount(CurrencyAmount.ether(exactInBribe), chainId)
+    if (etherIn){
+      amountOut = wrappedAmount(CurrencyAmount.ether(exactOutBribe), chainId)
+    } 
+
+    let minTokenAmountIn: CurrencyAmount | TokenAmount | undefined
+    let minTokenAmountOut: CurrencyAmount | TokenAmount | undefined
+
+    for (let i = 0; i < pairs.length; i++){
+      const pair = pairs[i]
+      // pair irrelevant
+      if (!pair.token0.equals(amountOut.token) && !pair.token1.equals(amountOut.token)) continue 
+      if (pair.reserve0.equalTo(ZERO) || pair.reserve1.equalTo(ZERO)) continue
+
+      try {
+        if (etherIn){
+          minTokenAmountIn = CurrencyAmount.ether(exactInBribe)
+          ;[minTokenAmountOut] = pair.getInputAmount(amountOut, Exchange.UNI)
```

Comment:
Just saw that you are using `Exchange.UNI` can we make `exchange` a parameter and call accordingly where we need this method?